packages/react-router/src/useBlocker.tsx

```diff
@@ -3,25 +3,93 @@ import { useRouter } from './useRouter'
 import type { BlockerFn } from '@tanstack/history'
 import type { ReactNode } from './route'
 
+type BlockerResolver = {
+  status: 'idle' | 'blocked'
+  proceed: () => void
+  reset: () => void
+}
+
+type BlockerOpts = {
+  blockerFn?: BlockerFn
+  condition?: boolean | any
+}
+
+export function useBlocker(blockerFnOrOpts?: BlockerOpts): BlockerResolver
+
+/**
+ * @deprecated Use the BlockerOpts object syntax instead
+ */
+export function useBlocker(
+  blockerFn?: BlockerFn,
+  condition?: boolean | any,
+): BlockerResolver
+
 export function useBlocker(
-  blockerFn: BlockerFn,
-  condition: boolean | any = true,
-): void {
+  blockerFnOrOpts?: BlockerFn | BlockerOpts,
+  condition?: boolean | any,
+): BlockerResolver {
+  const { blockerFn, blockerCondition } = blockerFnOrOpts
+    ? typeof blockerFnOrOpts === 'function'
+      ? { blockerFn: blockerFnOrOpts, blockerCondition: condition ?? true }
+      : {
+          blockerFn: blockerFnOrOpts.blockerFn,
+          blockerCondition: blockerFnOrOpts.condition ?? true,
+        }
+    : { blockerFn: undefined, blockerCondition: condition ?? true }
   const { history } = useRouter()
 
-  React.useEffect(() => {
-    if (!condition) return
-    return history.block(blockerFn)
+  const [resolver, setResolver] = React.useState<BlockerResolver>({
+    status: 'idle',
+    proceed: () => {},
+    reset: () => {},
   })
+
+  const createPromise = () =>
+    new Promise<boolean>((resolve) => {
+      setResolver({
+        status: 'idle',
+        proceed: () => resolve(true),
+        reset: () => resolve(false),
+      })
+    })
+
+  const [promise, setPromise] = React.useState(createPromise)
+
+  React.useEffect(() => {
+    const blockerFnComposed = async () => {
+      // If a function is provided, it takes precedence over the promise blocker
+      if (blockerFn) {
+        return await blockerFn()
+      }
+
+      setResolver((prev) => ({
+        ...prev,
+        status: 'blocked',
+      }))
+      const canNavigateAsync = await promise
+
+      setPromise(createPromise)
+
+      return canNavigateAsync
+    }
+
+    return !blockerCondition ? undefined : history.block(blockerFnComposed)
+  }, [blockerFn, blockerCondition, history, promise])
```

Comment:
@Balastrong, thank you for the great work!

I noticed that at line 56 you used a `useState` to hold the promise, even there are nothing to be blocked. This can lead to an unexpected bug in the development mode, where React will run the some hooks twice to detect potential development issues, meaning there is a risk that a new promise is created while the previous promise never resolved. 

And in fact, there are indeed bugged because of the async `blockerFnComposed` function in the effect referencing the "old" promise (tested in React 19 beta).

Therefore, I suggest to only create a promise when there are something really to be promised, here is the suggested changes:


```ts
  React.useEffect(() => {
    async function blockerFnComposed() {
      // If a function is provided, it takes precedence over the promise blocker
      if (blockerFn) {
        return await blockerFn();
      }

      const promise = new Promise((resolve) => {
        setResolver({
          status: "blocked",
          proceed: () => resolve(true),
          reset: () => resolve(false),
        });
      });

      const canNavigateAsync = await promise;
      setResolver({
        status: "idle",
        proceed: () => {},
        reset: () => {},
      });

      return canNavigateAsync;
    }

    return !blockerCondition ? undefined : history.block(blockerFnComposed);
  }, [blockerFn, blockerCondition, history]);
  ```

Reply:
@leoyli
can you please create a PR to fix this issue?
ideally with a test 😁